RHEcosystemAppEng/dbaas-operator
# Enhance DBaaSConnection instanceRef handling to skip provider CR creation if instanceID is not available

Currently, DBaaS operator would create a provider instance CR, after constructing the spec of the CR regardless the validation of instanceRef data succeeds or not. If the validation fails, the operator creates a provider CR even though its spec is not valid.

This PR enhances this logic by checking the instanceRef validation result first, and creating the provider CR only when needed.

## Verification Steps
1) Create a provider account.
2) Create a DBaaSInstance CR with this account that fails during the instance provisioning, eg, for mongodb, create an instance with an existing cluster name would fail.
3) Create a DBaaSConnection CR with this instance but without instanceID populated.

Expected result:
The instance CR has the status like:
Status:
  Conditions:
    Last Transition Time:  2022-09-16T19:03:48Z
    Message:               instance ID is not available
    Reason:                DBaaSInstanceNotAvailable
    Status:                False
    Type:                  ConnectionReady

and no mongodbatlasconnection object is createad.

## Type of change

- [X ] Bug fix (non-breaking change which fixes an issue)

## Checklist (Definition of Done)

- [X ] all relevant tests are passing
